File: pytest-tests/cuboid.py
from rectangle import Rectangle, Square


class Cuboid:
    """
    Represents a 3D cuboid with a rectangular base and height.
    Provides methods to compute volume and surface area.
    """
    def __init__(self, base: Rectangle, height: float):
        self.base = base
        self.height = height

    def volume(self) -> float:
        """Returns the volume of the cuboid."""
        return self.base.area() * self.height

    def surface_area(self) -> float:
        """Returns the surface area of the cuboid."""
        lw = self.base.area()
        lh = self.base.length * self.height
        wh = self.base.width * self.height
        return 2 * (lw + lh + wh)


class Cube(Cuboid):
    """
    Represents a 3D cube, which is a special type of cuboid
    where the length, width, and height are all equal.
    """
    def __init__(self, side_length: float):
        base = Square(side_length)
        super().__init__(base, side_length)

File: jest-tests/cuboid.js
const { Rectangle } = require('./rectangle')

/**
 * Represents a cuboid with a rectangular base and a given height.
 */
class Cuboid {
  /**
   * Creates a new Cuboid with a rectangular base.
   * @param {Rectangle} base - The rectangular base of the cuboid.
   * @param {number} height - The height of the cuboid.
   */
  constructor(base, height) {
    this.base = base
    this.height = height
  }

  /**
   * Calculates the volume of the cuboid.
   * @returns {number} The volume of the cuboid.
   */
  volume() {
    return this.base.area() * this.height
  }

  /**
   * Calculates the surface area of the cuboid.
   * @returns {number} The surface area of the cuboid.
   */
  surfaceArea() {
    const lw = this.base.area()
    const lh = this.base.length * this.height
    const wh = this.base.width * this.height
    return 2 * (lw + lh + wh)
  }
}

/**
 * Represents a cube, which is a special case of a cuboid with all sides equal.
 */
class Cube extends Cuboid {
  /**
   * Creates a new Cube with a given side length.
   * @param {number} sideLength - The length of each side of the cube.
   */
  constructor(sideLength) {
    super(new Rectangle(sideLength, sideLength), sideLength)
  }
}

module.exports = { Cuboid, Cube }
